Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,342 @@
+     # Defaults to `Classic`.
+    mode: 'Classic'    
+
+    # The following section is only valid if "mode" is set to "Topology"
+    # 'topology.kubernetes.io/zone' for 'topology'
+    topology: 
+        # A kubernetes node label used in the nodeSelector field of prometheus
+        # instances. This assures that instances are placed in the zone they
+        # are meant to be scraping.
+        # Defaults to `topology.kubernetes.io/zone`
+        nodeLabel: 'topology.kubernetes.io/zone'
+
+        # All topology values to be used by nodeLabel and sourceLabel
+        values: []
+```
+
+The `topology` section does not use the term `zone`. This makes the feature
+more flexible in case a user needs to shard on e.g. regions instead.
+
+Both modes do not contain an explicit overwrite of the label used for sharding.
+This feature is already possible by generating a `__tmp_hash` label through
+[scrape classes](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.ScrapeClass).
+
+In the `Classic` mode this allows overriding the main label used for sharding.
+In case of the `Topology` mode, a different label might need to be used for
+e.g. services or pods. This label is automatically generated.
+The logic used for the `Classic` mode is still in place here, to support a
+custom label when assigning multiple instances to the same zone.
+In this case the `Classic` mode is to be seen as a subset of the `Topology`
+mode.
+
+## Generated configuration
+
+The following examples are based on the algorithm above.
+Please note that `shard_index` has to be provided by the operator during
+config generation.
+
+We use a replica count of 2 in all examples to illustrate that this value
+does not have any effect, as both replicas will have the same `shared_index`
+assigned.
+
+### classic mode
+
+Given the following configuration
+
+```yaml
+spec:
+  shards: 4
+  replicas: 2
+  shardingStrategy:
+    mode: 'Classic'
+```
+
+we would get the following output for `shard_index == 2`
+
+```yaml
+- source_labels: ['__address__', '__tmp_hash']
+  target_label: '__tmp_hash'
+  regex: '(.+);'
+  replacement: '$1'
+  action: 'replace'
+- source_labels: ['__tmp_hash']
+  target_label: '__tmp_hash'
+  modulus: 4
+  action: 'hashmod'
+- source_labels: ['__tmp_hash']
+  regex: '2'                    # shard_index
+  action: 'keep'
+```
+
+### topology mode
+
+Given the following configuration
+
+```yaml
+spec:
+  shards: 4
+  replicas: 2
+  shardingStrategy:
+    mode: 'Topology'    
+    topology:
+      nodeLabel: 'topology.kubernetes.io/zone'
+      values:
+        - 'europe-west4-a'
+        - 'europe-west4-b'
+```
+
+we would get the following output for `shard_index == 2`:
+
+```yaml
+# zones := shardingStrategy.topology.values
+# shards_per_zone := max(1, floor(shards / len(zones)))
+- source_labels: ['__meta_kubernetes_endpointslice_endpoint_zone']
```

Comment:
I already mentioned this above, but we could actually add both, the `__meta_kubernetes_endpointslice_endpoint_zone` and the node metadata field, couldn't we?

Reply:
yes the operator can pick up the right label based on the following decision tree:
* Use `__meta_kubernetes_endpointslice_endpoint_zone` if the resource is a service monitor + `.spec.serviceDiscoveryRole: EndpointSlice`
* Otherwise use `__meta_kubernetes_node_label_topology_kubernetes_io_zone` + attach node metadata.

And if we want to offer flexibility for users, we can take the same approach as with the `__tmp_hash` label and have the `keep` rule checking against a specific temporary label:

```
- source_labels: ['__meta_kubernetes_endpointslice_endpoint_zone', __tmp_topology]
  target_label: __tmp_topology
  regex: "(.+);"
- source_labels: ['__meta_kubernetes_node_label_topology_kubernetes_io_zone', '__meta_kubernetes_node_labelpresent_topology_kubernetes_io_zone','__tmp_topology']
  regex: "(.+);true;"
  target_label: __tmp_topology
- source_labels: ['__tmp_topology']
  regex: 'europe-west4-a'          # zones[shard_index % shards_per_zone]
  action: 'keep'
```

Reply:
I added this to the example and improved the overwrite mechanics explanation.
@simonpasquier is this now better/clearer?